crates/transforms/src/function_dispatcher/patterns/controller.rs

```diff
@@ -0,0 +1,626 @@
+
+use azoth_core::decoder::Instruction;
+use azoth_core::Opcode;
+
+/// Configuration for byte extraction pattern in a controller block.
+///
+/// This pattern re-extracts a single byte from the calldata within the controller
+/// and compares it against a derived token value. This adds an additional layer
+/// of verification beyond the main dispatcher's 4-byte selector comparison.
+#[allow(dead_code)]
+#[derive(Clone, Debug)]
+pub struct ByteExtractionPattern {
+    /// The byte index to extract (typically 0-3 for selector bytes)
+    pub byte_index: u8,
+    /// The expected byte value to compare against
+    pub expected_value: u8,
+    /// Program counter where the pattern should be inserted
+    pub insert_pc: usize,
+}
+
+/// Configuration for storage-based routing pattern in a controller block.
+///
+/// This pattern checks storage slot values to determine routing decisions,
+/// making control flow dependent on contract state.
+#[allow(dead_code)]
+#[derive(Clone, Debug)]
+pub struct StorageRoutingPattern {
+    /// Storage slot to load
+    pub slot: u64,
+    /// Expected value in the storage slot
+    pub expected_value: u128,
+    /// Target PC if the condition matches
+    pub target_pc: usize,
+}
+
+/// Configuration for opaque predicate pattern in a controller block.
+///
+/// This pattern inserts conditionals that always evaluate to a known value
+/// but are difficult to statically analyze, creating dead code paths.
+#[allow(dead_code)]
+#[derive(Clone, Debug)]
+pub struct OpaquePredicatePattern {
+    /// The predicate type (e.g., "1 == 0" always false, "x | !x" always true)
+    pub predicate_type: PredicateType,
+    /// Target PC for the "true" branch (typically invalid/decoy)
+    pub true_target: usize,
+    /// Target PC for the "false" branch (typically real path)
+    pub false_target: usize,
+}
+
+#[allow(dead_code)]
+#[derive(Clone, Debug)]
+pub enum PredicateType {
+    /// Always false: 1 == 0
+    AlwaysFalse,
+    /// Always true: 0 == 0
+    AlwaysTrue,
+    /// More complex: (x | !x) == all_bits_set (always true)
+    BitwiseIdentity,
+}
+
+/// Generates instructions for a storage check pattern.
+///
+/// Creates a sequence that loads a value from a storage slot and checks if it
+/// equals zero (the default state). This pattern leverages the fact that uninitialized
+/// storage slots default to zero, creating functional routing without requiring SSTORE.
+///
+/// # Parameters
+///
+/// * `start_pc` - Starting program counter for the pattern
+/// * `slot` - Storage slot to check
+/// * `match_target` - Where to jump if the slot value is zero
+/// * `fallback_target` - Where to jump if the slot value is non-zero
+///
+/// # Returns
+///
+/// A tuple of (generated instructions, next available PC).
+pub fn generate_storage_check_instructions(
+    start_pc: usize,
+    slot: u64,
+    match_target: usize,
+    fallback_target: usize,
+) -> (Vec<Instruction>, usize) {
```

Comment:
There's a few places where we're returning the updated pc, after vec<instruction>. Might be nice to impl an extension trait like

```
impl EncodedSize for Instruction {
  fn size(&self) -> usize { 
      match self { ... }
  }
}

impl<V: EncodedSize, T: Iter<Item = V>> EncodedSize for T {
  fn size(&self) -> usize { 
      self.iter().map(|v| v.size()).sum()
  }
}
```

And then we can just instructions.size() wherever we need to compute pc offsets or increment after an added block